### PR TITLE
refactor: migrate to `8.0.0` of `level`

### PR DESCRIPTION
**Pull request by faustbrian**
> Follow-up to https://github.com/ethereumjs/ethereumjs-monorepo/pull/1917

**Changes**
- A few packages now depend on `level` for proper type hinting. The `level` package itself is bundled for server and client-side use. It'll load whatever it needs to use depending on the environment.
- As specified by their upgrade `import level from 'level';` is now `import { Level } from 'level';` and `level()` is replaced with `new Level()` because classes are now used.
- `level` now has its own types instead of having to rely on third-party packages for that
- Replaced `level-mem` with https://github.com/Level/memory-level
- Removed the `MemoryDB` implementation in favour of https://github.com/Level/memory-level **(Maybe `MemoryDB ` could be kept somewhere in docs as a reference implementation?)**
- Removed `level-ws` because it wasn't used anywhere
- Depend on https://github.com/vweevers/module-error to mirror the error behaviour of `abstract-level`
- Because we now use `level` and `memory-level` there are 3 different possible instances of `abstract-level` but they all have a consistent interface due to `abstract-level`. These instances are `classic-level`, `browser-level` and `memory-level`. This should now make it possible or at least a lot easier to use the packages in browsers without polyfills for `level`.

**References**
- https://github.com/Level/level/releases/tag/v8.0.0
- https://github.com/Level/level/blob/v8.0.0/UPGRADING.md#800

### CODE REVIEW

**Comment by holgerd77** [comment]
> @holgerd77 left a few comments here that would be good to get answers to before I make the final changes and fix tests

Hi there, thanks so much for all the work here!!! Nice!

Can you give a short summary (4-5 bullet points or so) what's the essence of the changes here? So what libraries needed to be added (and why)? What API/type usage changes?

That would be great to quickly grasp the broader picture. 🙂

**Comment by holgerd77** [comment]
> Added to the opening post.

Thanks a lot!

I'll leave it to @acolytec3 to give some comments here on the final questions, he is more of a Level DB expert within the team and should be awake soon! 😁

File: packages/trie/package.json
Comment (holgerd77) [reply packages/trie/package.json:44]: I would think so, guess you can remove.

File: packages/blockchain/src/db/manager.ts
Comment (holgerd77) [reply packages/blockchain/src/db/manager.ts:30]: I am not enough into Level DB stuff and I have to say that I have never really gotten this whole LevelUp vs LevelDown vs Level vs AbstractLevelDown thing, so I have to say that I am 100% out here.

File: packages/trie/docs/classes/BaseTrie.md
Comment (acolytec3) [reply packages/trie/docs/classes/BaseTrie.md:62]: General comment about anything in a package's `docs/...` folder is that you don't need to update those.  We have a `docs` that updates those based on the Typedocs and function signatures.

File: packages/ethash/src/index.ts
Comment (acolytec3) [reply packages/ethash/src/index.ts:153]: I'll respond here for all the questions about the `Level<any,any>` typing question. Is it more trouble than it's worth to type it according to how we use keys and values in the DB?  I'm asking because I don't know but we allow `string | Buffer` today in keys and `Buffer | object` in values following [this](https://github.com/ethereumjs/ethereumjs-monorepo/blob/8a65e1928bc555519d22d3a5d5f885a54eaf0eff/packages/blockchain/src/db/operation.ts#L32).  If you define the db as `Level<string | Buffer, Buffer | object>` do you find any issues?

File: packages/client/lib/util/metaDBManager.ts
Comment (acolytec3) [reply packages/client/lib/util/metaDBManager.ts:31]: Now that I'm looking at some of these more closely, is there a reason to not define it as `Level<Buffer, Buffer>`?

File: packages/client/lib/client.ts
Comment (acolytec3) [reply packages/client/lib/client.ts:23]: Unless there's some reason I'm not tracking, I would keep these typed the way they are.

File: packages/client/lib/client.ts
Comment (acolytec3) [reply packages/client/lib/client.ts:31]: Unless there's some reason I'm not tracking, I would keep these typed the way they are.

File: packages/ethash/src/index.ts
Comment (acolytec3) [reply packages/ethash/src/index.ts:153]: Okay, that doesn't feel terrible to me since it should be localized within the `DBManager` or the related places deep in the code where we do DB ops.  If we can keep the external interfaces that return values to the various consumers intact, I would still like to have the DB key and value types defined wherever possible.

File: packages/blockchain/src/db/manager.ts
Comment (acolytec3) [reply packages/blockchain/src/db/manager.ts:183]: @holgerd77 I can't imagine any circumstance where we would want a key with an undefined value to be set.  I'd rather throw in such instances.

File: packages/blockchain/src/db/manager.ts
Comment (acolytec3) [reply packages/blockchain/src/db/manager.ts:180]: I see what you mean and not sure why `//eslint-disable-next-line @typescript-eslint/await-thenable` isn't clearing the lint error.  I think you can partially solve it by doing `(await this._db.get<Buffer, Buffer>(dbKey, dbOpts)) as Buffer` as that will at least satisfy the type inference on the return value.  I'll have to do some more digging as to why eslint is misbehaving here even though we tell it to.

File: packages/client/lib/util/metaDBManager.ts
Comment (acolytec3) [reply packages/client/lib/util/metaDBManager.ts:31]: I don't mind that at all.  LGTM.

File: packages/blockchain/src/db/manager.ts
Comment (acolytec3) [reply packages/blockchain/src/db/manager.ts:181]: Yes, the retrieved values should be all parsed by consuming methods.

**Comment by holgerd77** [comment]
> Looks like there's some issue with browser builds and also some odd ones with `test-client-cli` that have no clear error message. Need to dig into those.

There are some *known* browser build issues on the client. So if you refer to these you can safely ignore. Also the client CLI tests (this is basically life-running a client on a network and see if blocks come in) are not 100% reliable, you can also mostly ignore those test failures.

**Comment by holgerd77** [comment]
> Should retrying eventually get those tests to pass? Will look into the other failing tests first.

No, test failures are totally unrelated. The other tests we need to fix though.

File: packages/trie/docs/classes/BaseTrie.md
Comment (acolytec3) [reply packages/trie/docs/classes/BaseTrie.md]: I think we can do it separately as part of our next release.  @holgerd77 is that accurate?

**Comment by acolytec3** [comment]
Is the `ModuleError` dependency necessary?  I may have missed it in the code but do we have to have this dependency in order to catch and appropriately handle errors the `level` errors correctly?  I'm not tracking how this helps us if we are catching errors within the DBmanager and not exposing them directly to application consumers.

File: packages/ethash/src/index.ts
Comment (acolytec3) [inline packages/ethash/src/index.ts:143]: Is `level-mem` not compliant with the `abstract-level` API?  I see that you're typing it to `any` in a number of places and that typescript is complaining about a missing `location` property if you don't do so.

File: packages/ethash/examples/rawExample.ts
Comment (acolytec3) [inline packages/ethash/examples/rawExample.ts:4]: Can you cast `cacheDB` to any here if `memory-level` doesn't provide the exact interface conforming to `abstract-level`?

File: packages/ethash/src/index.ts
Comment (acolytec3) [reply packages/ethash/src/index.ts:143]: That makes sense to me.  This is a big improvement overall so I'm fine with that for now since additional fixes would definitely not be breaking changes for us.

**Comment by acolytec3** [comment]
> @acolytec3 can you check [0f971fa](https://github.com/ethereumjs/ethereumjs-monorepo/commit/0f971fab4e4deadf70f2cbf3ca529bedf3c4956c)

This looks great!  Can you address my question about the necessity of using the `ModuleError` dependency?

**Comment by acolytec3** [comment]
> Do you have a link to it? Can't see it anywhere and no notification 🤔

https://github.com/ethereumjs/ethereumjs-monorepo/pull/1949#issuecomment-1152894941

**Comment by acolytec3** [comment]
> > Is the `ModuleError` dependency necessary? I may have missed it in the code but do we have to have this dependency in order to catch and appropriately handle errors the `level` errors correctly? I'm not tracking how this helps us if we are catching errors within the DBmanager and not exposing them directly to application consumers.
> 
> Previously we threw a level-specific exception at https://github.com/ethereumjs/ethereumjs-monorepo/pull/1949/files#diff-f8b3e1c1192edac8a59bf9b474632a9e21e259fd79705b3a1bb416391f66fa72L154 and the new behaviour in `abstract-level` is to throw errors with `ModuleError` so using this here makes sure that we reflect the new way in which `level` throws errors. Changing this caused quite a few places to blow up but I can take another look if wanted.

Got it. Let me experiment with it locally and think about the change.

**Comment by acolytec3** [comment]
How hard would it be to just pull the `ModuleError' class on from the dependency and use it in our code? It's a small dependency from looking at it and no reason we can't just adopt it internally. That way we get the advantage of the level specific errors without the actual dependency.

**Comment by acolytec3** [comment]
> > How hard would it be to just pull the `ModuleError' class on from the dependency and use it in our code? It's a small dependency from looking at it and no reason we can't just adopt it internally. That way we get the advantage of the level specific errors without the actual dependency.
> 
> Will take care of that in a few hours +1

If you're open to friendly commits, I can go ahead and push it.

File: packages/client/test/sync/skeleton.spec.ts
Comment (acolytec3) [inline packages/client/test/sync/skeleton.spec.ts:162]: Can't we remove these `any` type casts now as well since you changed the types over to `abstract-level`? I checked locally and Typescript seems to be happy with it without the `:any` casting.

Comment (acolytec3) [reply packages/client/test/sync/skeleton.spec.ts:162]: This is my last nit.  This is fantastic work so thank you so much!  I know this has been a long slog.

**Comment by acolytec3** [comment]
@holgerd77 I'm ready to approve this once the tests pass again.  Will leave it to you to make a final call on `wen merge` as it were but this looks great to me.

**Comment by acolytec3** [comment]
> I'll open a clean PR within 30 minutes or so unless it's fine as is? I remember last time it had to be rebased before squashing.

Github isn't complaining about any discrepancies with current master so no reason to rebase again at this point.  @holgerd77, if you're good with this, we should go ahead and merge before any more work causes another massive headache of rebasing (thinking things like #1950)

**Comment by acolytec3** [comment]
> Not sure why https://github.com/ethereumjs/ethereumjs-monorepo/runs/6846930504?check_suite_focus=true is having such a massive performance regression after the last commit. Random failure?

Likely, our benchmarks are dubious at best and generate warnings for very seemingly unrelated code changes.

**Review by holgerd77 — APPROVED** [review]
Apart from these insane change sets we are merging in last minute:

This looks really great, thanks a lot @faustbrian for this yet-another-round of improvements! 🎉

This will (should) nevertheless likely be a candidate to have some closer look and experiment with during the beta release period, since this is so new and really affects relatively prominent parts of our interfaces.

Neverthess: cool, great work!

Will merge.

**Comment by holgerd77** [comment]
(and - yes - the strict rebase policy was solely for the `develop` work, rebase still preferred though if possible/practicable)

**Pull request merged**
